### PR TITLE
MEN-4074: Setup signal handler earlier to avoid unhandled SIGUSR sign…

### PR DESCRIPTION
**Pull request by NilsOlavKJohansen**
…als.

Changelog: None
Signed-off-by: Nils Olav Kvelvane Johansen <nils.olav@northern.tech>

# External Contributor Checklist

🚨 Please review the [guidelines for contributing](https://github.com/mendersoftware/mender/blob/master/CONTRIBUTING.md) to this repository.

- [x] Make sure that all commits have a [`Changelog`](https://github.com/mendersoftware/mender/blob/master/CONTRIBUTING.md#changelog-tags) tag. If nothing should be added to the Changelog, add a `Changelog: None` tag. If there is a change, add `Changelog: Commit`, or `Changelog: Title`, depending on what should be included in the changelog.

- [x] Make sure that all commits are signed with [`git --signoff`](https://git-scm.com/book/en/v2/Git-Tools-Signing-Your-Work). Also note that the signoff author must match the author of the commit.

### Description

Please describe your pull request.

Thank you!

### CODE REVIEW

**Comment by NilsOlavKJohansen** [comment]
@mender-test-bot start pipeline

**Review by kacf — CHANGES_REQUESTED** [review]
This should have a changelog, since it fixes a publicly visible bug. Think "what changed from the user's perspective?" when writing changelog entries. It's a good idea to read [this section](https://github.com/mendersoftware/mender/blob/master/CONTRIBUTING.md#changelog-tags) if you haven't already, to get it under your skin.

Other than that, I think the code looks good, but @oleorhagen should have a look too, since he's more familiar with this code than me.

**Comment by oleorhagen** [comment]
You figured it out @NilsOlavKJohansen! Good job!

The implementation is fine as it is IMO, but we could possible make it even better I think.

As I see it we have two approaches:
1. Sort of what you have done now, with a global signal handler channel. But then I think we could also take advantage of an `init()` function, to set this up, as soon as possible. I might actually wager that this would then fit in `main.go` even, so that it is actually the first thing which happens when running our binary, and hence, there is no window left for crashing the binary upon sending it a signal too soon.
2. If we're not doing approach nr 1. above, then I see no need for a global channel, and it might as well be sent as an argument to the daemon function.

What do you think?

**Comment by NilsOlavKJohansen** [comment]
> This should have a changelog, since it fixes a publicly visible bug. Think "what changed from the user's perspective?" when writing changelog entries. It's a good idea to read [this section](https://github.com/mendersoftware/mender/blob/master/CONTRIBUTING.md#changelog-tags) if you haven't already, to get it under your skin.

Thanks Kristian! I'll have a look at it :+1:

**Comment by NilsOlavKJohansen** [comment]
> You figured it out @NilsOlavKJohansen! Good job!
> 
> The implementation is fine as it is IMO, but we could possible make it even better I think.
> 
> As I see it we have two approaches:
> 
>     1. Sort of what you have done now, with a global signal handler channel. But then I think we could also take advantage of an `init()` function, to set this up, as soon as possible. I might actually wager that this would then fit in `main.go` even, so that it is actually the first thing which happens when running our binary, and hence, there is no window left for crashing the binary upon sending it a signal too soon.
> 
>     2. If we're not doing approach nr 1. above, then I see no need for a global channel, and it might as well be sent as an argument to the daemon function.
> 
> 
> What do you think?

I think I prefer 1. in main.go. As it eliminates the chance of this crashing the client to the greatest extent.

**Comment by NilsOlavKJohansen** [comment]
@mender-test-bot start-pipeline

**Comment by NilsOlavKJohansen** [comment]
@mender-test-bot start pipeline

File: cli/commands.go [hunk 1]
```diff
@@ -305,18 +305,17 @@ func PrintProvides(device *dev.DeviceManager) error {
 	return nil
 }
 
+//The channel gets initialized at boot
+var SignalHandlerChan chan os.Signal
```
Comment (NilsOlavKJohansen) [inline cli/commands.go:309]: I understand that SignalHandlerChan here in commands needs to Capitalized "C" to be exported. But should I follow the standard camel case on the local SignalHandlerChan varables, like on line 142 in cli_test.go and 29 in main.go. Or should I have all capitalized "C" to be consistent?

File: cli/commands.go [hunk 2]
```diff
@@ -66,6 +66,9 @@ var (
 	ErrSIGTERM           = errors.New("Daemon terminated with SIGTERM")
 )
 
+//The channel gets initialized at boot
+var SignalHandlerChan chan os.Signal
```
Comment (NilsOlavKJohansen) [inline cli/commands.go:70]: I understand that SignalHandlerChan in commands needs to Capitalized "C" to be exported. But should I follow the standard camel case on the local SignalHandlerChan varables. for example the variable on line 29 in main.go. Or should I have it the way it is now to be consistent?

Comment (oleorhagen) [reply cli/commands.go:70]: No, you don't. There is no need to export it from main :)

**Comment by NilsOlavKJohansen** [comment]
@mender-test-bot start pipeline

**Comment by oleorhagen** [comment]
@mender-test-bot start pipeline

**Comment by oleorhagen** [comment]
@mender-test-bot start build

**Comment by oleorhagen** [comment]
@mender-test-bot start build

**Comment by oleorhagen** [comment]
@mender-test-bot start pipeline

**Review by oleorhagen — APPROVED** [review]

File: cli/cli_test.go [hunk 3]
```diff
@@ -137,6 +138,8 @@ func TestRunDaemon(t *testing.T) {
 			ForceToState: make(chan app.State, 1),
 		}
 		go func() {
+			SignalHandlerChan = make(chan os.Signal, 2)
```
Comment (oleorhagen) [inline cli/cli_test.go:141]: The signalhandler channel is intialized twice.
Once here

File: cli/commands.go [hunk 4]
```diff
@@ -66,6 +66,8 @@ var (
 	ErrSIGTERM           = errors.New("Daemon terminated with SIGTERM")
 )
 
+var SignalHandlerChan = make(chan os.Signal, 2)
```
Comment (oleorhagen) [inline cli/commands.go:69]: and once here

**Review by oleorhagen — CHANGES_REQUESTED** [review]
See the comments :)

**Pull request merged**